src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -110,6 +109,18 @@ private void TryOpeningDocumentsForMonikerAndSetContextOnUIThread(string moniker
                 });
             }
 
+            private async Task TryOpeningDocumentsForMonikerAndSetContextOnUIThreadAsync(string moniker, ITextBuffer textBuffer, IVsHierarchy? hierarchy, CancellationToken cancellationToken)
+            {
+                await _projectSystemProjectFactory.ApplyChangeToWorkspaceAsync(async w =>
+                {
+                    await _foregroundAffinitization.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
This switch here while holding the lock makes me a bit nervous -- even if we won't deadlock here, it still means if the UI thread is unavailable we're holding up the lock and preventing other stuff from making progress....

Reply:
This switch would not be necessary if `ApplyChangeToWorkspaceAsync` used `ConfigureAwait(true)` on the following line:
https://github.com/dotnet/roslyn/blob/d11caad0ab35ec679716353eead320f92d05e753/src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs#L200

The current implementation is making assumptions about the caller that do not hold in practice.

Reply:
There is a synchronous wait via ApplyChangeToWorkspace that causes a deadlock here: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1920741. Please turn this into a JTF-aware lock, or delete all the async overloads which make it very tempting to try to switch to the UI thread.